### PR TITLE
Fix eip2929 implementation

### PR DESCRIPTION
**Pull request by RomarQ**
Includes a fix related to [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929), where some addresses were incorrectly considered `cold` and consumed extra gas. I identified this issue when running the tests located at [`ethtests/GeneralStateTests/Cancun/stEIP1153-transientStorage`](https://github.com/ethereum/tests/tree/develop/src/GeneralStateTestsFiller/Cancun/stEIP1153-transientStorage), which were failing because the gas costs did not match.

### CODE REVIEW

**Comment by RomarQ** [comment]
@sorpaas can you have a look at these changes?

I splitted these changes from the #278 to make the review easier.

**Comment by koushiro** [comment]
@RomarQ could you split this PR into three PRs?
- PR1: fix eip1559
- PR2: fix eip2929 (and update jsontests in CI)
- PR3: add cancun config

**Review by koushiro — APPROVED** [review]

**Comment by koushiro** [comment]
Could you add [`ethtests/GeneralStateTests/Cancun/stEIP1153-transientStorage`](https://github.com/ethereum/tests/tree/develop/src/GeneralStateTestsFiller/Cancun/stEIP1153-transientStorage) into jsontests in CI

**Comment by RomarQ** [comment]
> Could you add [`ethtests/GeneralStateTests/Cancun/stEIP1153-transientStorage`](https://github.com/ethereum/tests/tree/develop/src/GeneralStateTestsFiller/Cancun/stEIP1153-transientStorage) into jsontests in CI

This depends on https://github.com/rust-ethereum/evm/pull/288. 

We can test it this way:
```sh
cargo run --release --verbose -p jsontests -- jsontests/res/ethtests/GeneralStateTests/Cancun/stEIP1153-transientStorage --debug
```

**Comment by RomarQ** [comment]
Once we enable tests for `Cancun` fork here:
https://github.com/rust-ethereum/evm/blob/8168ca150abce577a4fb78d2da0918d229939b9e/jsontests/src/run.rs#L106

if we run all the tests, some will fail, mainly because there are still a few incompatibilities after berlin (gas related). I only fixed a few, there are more.

**Comment by koushiro** [comment]
> Once we enable tests for `Cancun` fork here:
> 
> https://github.com/rust-ethereum/evm/blob/8168ca150abce577a4fb78d2da0918d229939b9e/jsontests/src/run.rs#L106
> 
> if we run all the tests, some will fail, mainly because there are still a few incompatibilities after berlin (gas related). I only fixed a few, there are more.

We could fix those incompatibilities in future PRs, let's merge this PR first

File: src/standard/gasometer/mod.rs [hunk 1]
```diff
@@ -307,40 +307,59 @@ fn dynamic_opcode_cost<H: RuntimeBackend>(
 
 		Opcode::EXTCODESIZE => {
 			let target = stack.peek(0)?.into();
-			GasCost::ExtCodeSize {
-				target_is_cold: handler.is_cold(target, None),
-			}
+
+			// https://eips.ethereum.org/EIPS/eip-2929
+			let target_is_cold = handler.is_cold(target, None);
+			handler.mark_hot(target, None);
```
[the remark this answers is not in the document]
Comment (RomarQ) [reply src/standard/gasometer/mod.rs:313]: Yes, in some circumstances, the addresses were not being considered as `warm` when they should. If the objective is to have the code as customisable as possible, probably having it in the gasometer is not that bad.

**Pull request merged**